java-checks/src/main/java/org/sonar/java/checks/StringConcatToTextBlockCheck.java

```diff
@@ -0,0 +1,109 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Stream;
+import org.sonar.check.Rule;
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S6126")
+public class StringConcatToTextBlockCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  private static final String MESSAGE = "Replace this String concatenation with Text block.";
+  public static final int MINIMAL_LINE_LENGTH = 7;
+  public static final int MINIMAL_NUMBER_OF_LINES = 2;
+  private final Set<Tree> visitedNodes = new HashSet<>();
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava15Compatible();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.PLUS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (visitedNodes.contains(tree)) {
+      return;
+    }
+    List<Tree> concatOperands = getConcatOperands(((BinaryExpressionTree) tree));
+    
+    // We cannot report issues if there are non literal operands in concatenation
+    if (concatOperands.stream().anyMatch(operand -> !(operand instanceof LiteralTree))) {
+      return;
+    }
+    
+    long outputLines = concatOperands.stream()
+      .filter(op -> op.is(Tree.Kind.STRING_LITERAL))
+      .map(LiteralTree.class::cast)
+      .map(LiteralTree::value)
+      .filter(value -> value.contains("\\n"))
+      .flatMap(value -> Stream.of(value.split("\\\\n")))
+      .filter(str -> str.length() >= MINIMAL_LINE_LENGTH)
+      .count();
+
+    if (outputLines >= MINIMAL_NUMBER_OF_LINES) {
+      reportIssue(tree, MESSAGE);
+    }
```

Comment:
I like the result of this implementation but I would like to propose another approach.

1. I want to split the lines of the full concatenated string content and not some fragment of it to have more consistent findings.
```
    String e = "abcdef\n" + "abcdef\n"; // currently non-compliant
    String f = "abcdef\nabc" + "def\n"; // currently compliant
```

2. I don't want to take into account the double quotes delimiter in the computation
```
    String h = "123456\n1234567\n" + "\n"; // currently non-compliant, it's <<"123456>> and <<1234567>>
    String g = "1234567\n123456\n" + "\n"; // currently compliant, it's <<"1234567>> and <<123456>>
```

3. It should be easy to ignore when it's not a new line `"\n"` but two distinct characters `"\\n"`
 
4. it should be possible to get rid of `Set<Tree> visitedNodes` by looking at the parent node kind.

So here is my proposal:
```
  public static final int MINIMAL_CONTENT_LENGTH = 19;
  public static final int MINIMAL_NUMBER_OF_NEW_LINES = 2;
  // ...
  @Override
  public void visitNode(Tree tree) {
    if (isChildOfPlusOperator(tree)) {
      return;
    }
    StringBuilder concatenatedContent = new StringBuilder();
    if (concatStringLiterals(concatenatedContent, tree)) {
      String content = concatenatedContent.toString();
      if (content.length() >= MINIMAL_CONTENT_LENGTH &&
        !content.contains("\\\\n") &&
        StringUtils.countMatches(content, "\\n") >= MINIMAL_NUMBER_OF_NEW_LINES) {
        reportIssue(tree, MESSAGE);
      }
    }
  }

  private static boolean isChildOfPlusOperator(Tree child) {
    Tree parent = child.parent();
    while (parent.is(Tree.Kind.PARENTHESIZED_EXPRESSION)) {
      parent = parent.parent();
    }
    return parent.is(Tree.Kind.PLUS);
  }

  private static boolean concatStringLiterals(StringBuilder concatenatedContent, Tree tree) {
    if (tree.is(Tree.Kind.PLUS)) {
      BinaryExpressionTree binaryExpression = (BinaryExpressionTree) tree;
      return concatStringLiterals(concatenatedContent, ExpressionUtils.skipParentheses(binaryExpression.leftOperand())) &&
             concatStringLiterals(concatenatedContent, ExpressionUtils.skipParentheses(binaryExpression.rightOperand()));
    } else if (tree instanceof LiteralTree) {
      String treeValue = ((LiteralTree) tree).value();
      if (tree.is(Tree.Kind.STRING_LITERAL)) {
        treeValue = LiteralUtils.trimQuotes(treeValue);
      }
      concatenatedContent.append(treeValue);
      return true;
    } else {
      return false;
    }
  }
```

WDYT?

Reply:
I liked the idea of concatStringLiterals method. I changed it a little.
As for isChildOfPlusOperator() method. It's kinda questionable what's better, caching visited nodes, or calculating. I'd rather leave caching here.